scratchml/models/svc.py

```diff
@@ -0,0 +1,202 @@
+from scratchml.utils import split_data_into_batches
+from ..metrics import accuracy, recall, precision, f1_score, confusion_matrix
+from scratchml.regularizations import l2
+from scratchml.scalers import StandardScaler
+import numpy as np
+
+
+class SVC:
+    """
+    Support Vector Classifier (SVC) with options for linear, polynomial, and RBF kernels.
+    """
+
+    _valid_metrics = ["accuracy", "precision", "recall", "f1_score", "confusion_matrix"]
+
+    def __init__(
+        self,
+        C=0.35,
+        alpha=0.01,
+        kernel="linear",
+        degree=3,
+        max_iter=1500,
+        tol=1e-4,
+        learning_rate=1e-5,
+        decay=0.999,
+        batch_size=64,
+        early_stopping=True,
+        adaptive_lr=False,
+    ):
+        if C <= 0:
+            raise ValueError("Regularization parameter C must be positive.")
+
+        self.C = C
+        self.alpha = alpha
+        self.kernel = kernel
+        self.degree = degree
+        self.max_iter = max_iter
+        self.tol = tol
+        self.learning_rate = learning_rate
+        self.decay = decay
+        self.batch_size = batch_size
+        self.early_stopping = early_stopping
+        self.adaptive_lr = adaptive_lr
+        self.weights = None
+        self.classes_ = None
+        self.bias = 0
+        self.scaler = StandardScaler()
+        self._validate_parameters()
+
+    def _apply_kernel(self, X1, X2):
+        """Applies the selected kernel function."""
+        if self.kernel == "linear":
+            return np.dot(X1, X2.T)
+        elif self.kernel == "polynomial":
+            return (1 + np.dot(X1, X2.T)) ** self.degree
+        elif self.kernel == "rbf":
+            gamma = 1 / X1.shape[1]
+            return np.exp(
+                -gamma * np.square(np.linalg.norm(X1[:, np.newaxis] - X2, axis=2))
+            )
+
+    def _validate_parameters(self):
+        if self.C <= 0:
+            raise ValueError("Regularization parameter C must be positive.")
+
+    def _check_is_fitted(self):
+        if self.weights is None:
+            raise ValueError("Model must be trained before prediction.")
+
+    def fit(self, X, y):
+        """
+        Trains the SVC model using mini-batch gradient descent.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            Training data.
+        y : ndarray, shape (n_samples,)
+            Target labels (e.g., 0 or 1 for binary classification).
+        """
+        self.classes_ = np.unique(y)
+        if len(self.classes_) == 2:
+            y_ = np.where(y == self.classes_[0], -1, 1)
+        else:
+            y_ = y
+
+        X = self.scaler.fit_transform(X)
+        n_samples, n_features = X.shape
+        self.weights = np.zeros(n_features)
+
+        no_improvement_count = 0
+        for iteration in range(self.max_iter):
```

Comment:
I think it would be a good idea to put some comments throughout the `fit` function on some steps that you think it's important to understand the logic behind the code and the main idea of how the algorithm works.